// @storybook/addon-knobs/dist/components/types/Select.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint no-underscore-dangle: 0 */

var styles = {
  display: 'table-cell',
  boxSizing: 'border-box',
  verticalAlign: 'middle',
  height: '26px',
  width: '100%',
  outline: 'none',
  border: '1px solid #f7f4f4',
  borderRadius: 2,
  fontSize: 11,
  padding: '5px',
  color: '#555'
};

var SelectType = function (_React$Component) {
  (0, _inherits3.default)(SelectType, _React$Component);

  function SelectType() {
    (0, _classCallCheck3.default)(this, SelectType);
    return (0, _possibleConstructorReturn3.default)(this, (SelectType.__proto__ || (0, _getPrototypeOf2.default)(SelectType)).apply(this, arguments));
  }

  (0, _createClass3.default)(SelectType, [{
    key: '_makeOpt',
    value: function _makeOpt(key, value, selectV2) {
      var opts = {
        key: key,
        value: key
      };

      var display = value;

      if (selectV2) {
        opts.value = value;
        display = key;
      }

      return _react2.default.createElement(
        'option',
        opts,
        display
      );
    }
  }, {
    key: '_options',
    value: function _options(_ref) {
      var _this2 = this;

      var options = _ref.options,
          selectV2 = _ref.selectV2;

      var data = [];
      if (Array.isArray(options)) {
        data = options.map(function (val) {
          return _this2._makeOpt(val, val);
        });
      } else {
        data = (0, _keys2.default)(options).map(function (key) {
          return _this2._makeOpt(key, options[key], selectV2);
        });
      }

      return data;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          knob = _props.knob,
          _onChange = _props.onChange;


      return _react2.default.createElement(
        'select',
        {
          id: knob.name,
          ref: function ref(c) {
            _this3.input = c;
          },
          style: styles,
          value: knob.value,
          onChange: function onChange(e) {
            return _onChange(e.target.value);
          }
        },
        this._options(knob)
      );
    }
  }]);
  return SelectType;
}(_react2.default.Component);

SelectType.defaultProps = {
  knob: {},
  onChange: function onChange(value) {
    return value;
  }
};

SelectType.propTypes = {
  knob: _propTypes2.default.shape({
    name: _propTypes2.default.string,
    value: _propTypes2.default.string,
    options: _propTypes2.default.oneOfType([_propTypes2.default.array, _propTypes2.default.object]),
    selectV2: _propTypes2.default.bool
  }),
  onChange: _propTypes2.default.func
};

SelectType.serialize = function (value) {
  return value;
};
SelectType.deserialize = function (value) {
  return value;
};

exports.default = SelectType;
