// @storybook/addon-links/dist/index.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkTo = LinkTo;
Object.defineProperty(exports, "ADDON_ID", {
  enumerable: true,
  get: function get() {
    return _events.ADDON_ID;
  }
});
Object.defineProperty(exports, "EVENT_ID", {
  enumerable: true,
  get: function get() {
    return _events.EVENT_ID;
  }
});
Object.defineProperty(exports, "REQUEST_HREF_EVENT_ID", {
  enumerable: true,
  get: function get() {
    return _events.REQUEST_HREF_EVENT_ID;
  }
});
Object.defineProperty(exports, "RECEIVE_HREF_EVENT_ID", {
  enumerable: true,
  get: function get() {
    return _events.RECEIVE_HREF_EVENT_ID;
  }
});
Object.defineProperty(exports, "linkTo", {
  enumerable: true,
  get: function get() {
    return _preview.linkTo;
  }
});
Object.defineProperty(exports, "hrefTo", {
  enumerable: true,
  get: function get() {
    return _preview.hrefTo;
  }
});
Object.defineProperty(exports, "withLinks", {
  enumerable: true,
  get: function get() {
    return _preview.withLinks;
  }
});

var _events = require("./events");

var _preview = require("./preview");

var hasWarned = false;

function LinkTo() {
  if (!hasWarned) {
    // eslint-disable-next-line no-console
    console.error("\nLinkTo has moved to addon-links/react:\n\nimport LinkTo from '@storybook/addon-links/react';\n    ");
    hasWarned = true;
  }

  return null;
}
